src/sinks/elasticsearch/encoder.rs

```diff
@@ -108,35 +112,92 @@ fn write_bulk_action(
     doc_type: &str,
     suppress_type: bool,
     id: &Option<String>,
+    version: &Option<u64>,
+    version_type: &VersionType,
 ) -> std::io::Result<usize> {
     as_tracked_write(
         writer,
-        (bulk_action, index, doc_type, id, suppress_type),
-        |writer, (bulk_action, index, doc_type, id, suppress_type)| match (id, suppress_type) {
-            (Some(id), true) => {
+        (
+            bulk_action,
+            index,
+            doc_type,
+            id,
+            suppress_type,
+            version,
+            version_type,
+        ),
+        |writer, (bulk_action, index, doc_type, id, suppress_type, version, version_type)| match (
+            id,
+            suppress_type,
+            (version_type, version),
+        ) {
+            (_, _, (VersionType::External, None) | (VersionType::ExternalGte, None)) => {
+                Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Tried to use external versioning without specifying the version itself",
+                ))
+            }
+            (None, _, (VersionType::External, Some(_)) | (VersionType::ExternalGte, Some(_))) => {
+                Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "Cannot use external versioning without specifying a document ID",
+                ))
```

Comment:
These seem like errors we could detect when we first build the component, since they're purely from configuration. I would recommend that we do that validation there instead, at least as much as we can without rendering the template. That way we can immediately return a validation error rather than waiting until we actually try to write something. Also, we can probably simplify the version type after validating it and avoid complicating this already-complex piece of code here.